[SYCL] Add ballot_group support to algorithms

Pennycook opened the pull request:
Enables the following functions to be used with ballot_group arguments:
- group_barrier
- group_broadcast
- any_of_group
- all_of_group
- none_of_group
- reduce_over_group
- exclusive_scan_over_group
- inclusive_scan_over_group

Signed-off-by: John Pennycook <john.pennycook@intel.com>

Pennycook commented:
A few quick notes to reviewers:

1. I split this out into three commits to try and make the review a little easier. But I reconstructed them from the true history, so I'm not 100% sure if things will work if you try to build one of the earlier commits. :smile:

2. Tests are in https://github.com/intel/llvm-test-suite/pull/1698.

3. This may look like a lot of refactoring for the sake of one group type, but there are a few more coming.  The idea of having something like `spirv::GroupBroadcast` resolve either to `__spirv_GroupBroadcast` or a sequence of `__spirv_GroupNonUniformBroadcast` calls depending on the type of group is to facilitate simpler implementation of the remaining non-uniform groups.

JackAKirk commented on sycl/include/sycl/ext/oneapi/experimental/non_uniform_groups.hpp:
```diff
@@ -9,53 +9,63 @@
+      }
+    }
+  }
+  __builtin_unreachable();
```
Should this return an error? e.g. Id out of bounds/invalid etc?

Pennycook replied:
I don't think so.  This is device code, so we can't really throw a useful error.  The only way I can think of to signal an error would be to return some sort of sentinel value (e.g. always accessing lane 0 in the case of out-of-bounds behavior).

I thought that SYCL 2020 said this was UB, but it looks like that's not the case -- [the definition](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_group_broadcast) of `group_broadcast` doesn't say anything about the range of acceptable local IDs.  I think this is a bug, though, because the restrictions are present for the "shuffles".  I'll open an issue against the SYCL spec.

Pennycook replied:
Opened https://github.com/KhronosGroup/SYCL-Docs/pull/390.

Pennycook commented on sycl/include/sycl/detail/spirv.hpp:
```diff
@@ -94,13 +103,35 @@ void GenericCall(const Functor &ApplyToBytes) {
+              bool pred) {
+  // Each ballot_group implicitly represents two groups
+  // We have to force each half down different control flow
+  if (g.get_group_id() == 1) {
```
Yeah, exactly.

If you look at the definitions of some of these NonUniform instructions in SPIR-V (e.g. [`OpGroupNonUniformAll`](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformAll)), you can see that they're defined in terms of "active invocations". "active" isn't actually defined anywhere (see [this Khronos issue](https://gitlab.khronos.org/spirv/SPIR-V/-/issues/691)) but it's tied to control flow.

Pennycook commented:
@intel/llvm-reviewers-runtime, @steffenlarsen: Tests are passing, so this should be ready for review!

aelovikov-intel commented:
Hi @Pennycook , with intel/llvm-test-suite moved in-tree you need to add the tests from https://github.com/intel/llvm-test-suite/pull/1698 into this PR now.

Pennycook commented:
> Hi @Pennycook , with intel/llvm-test-suite moved in-tree you need to add the tests from [intel/llvm-test-suite#1698](https://github.com/intel/llvm-test-suite/pull/1698) into this PR now.

Thanks, @aelovikov-intel.  I've copied over the tests from https://github.com/intel/llvm-test-suite/pull/1698, and also the related tests from https://github.com/intel/llvm-test-suite/pull/1574 which didn't get merged before the move.

aelovikov-intel commented on sycl/include/sycl/detail/spirv.hpp:
```diff
@@ -208,17 +267,26 @@ EnableIfNativeBroadcast<T> GroupBroadcast(T x, id<Dimensions> local_id) {
+EnableIfNativeBroadcast<T>
+GroupBroadcast(sycl::ext::oneapi::experimental::ballot_group<ParentGroup> g,
+               T x, id<1> local_id) {
+  // Limited to 1D indices for now because ParentGroup must be sub-group
```
```suggestion
  // Limited to 1D indices for now because ParentGroup must be sub-group.
```

aelovikov-intel commented on sycl/test-e2e/NonUniformGroups/ballot_group_algorithms.cpp:
```diff
@@ -0,0 +1,131 @@
+          auto WI = item.get_global_id();
+          auto SG = item.get_sub_group();
+
+          // Split into odd and even work-items
```
```suggestion
          // Split into odd and even work-items.
```

aelovikov-intel commented on sycl/test-e2e/NonUniformGroups/ballot_group_algorithms.cpp:
```diff
@@ -0,0 +1,131 @@
+          bool Predicate = WI % 2 == 0;
+          auto BallotGroup = syclex::get_ballot_group(SG, Predicate);
+
+          // Check all other members' writes are visible after a barrier
```
```suggestion
          // Check all other members' writes are visible after a barrier.
```

aelovikov-intel commented on sycl/test-e2e/NonUniformGroups/ballot_group_algorithms.cpp:
```diff
@@ -0,0 +1,131 @@
+          }
+          BarrierAcc[WI] = Visible;
+
+          // Simple check of group algorithms
```
```suggestion
          // Simple check of group algorithms.
```

aelovikov-intel commented on sycl/test-e2e/NonUniformGroups/ballot_group_algorithms.cpp:
```diff
@@ -0,0 +1,131 @@
+            BroadcastAcc[WI] = (BroadcastResult == 1);
+          }
+
+          bool AnyResult = sycl::any_of_group(BallotGroup, Predicate);
```
I think it would make sense to change to a condition that is true for a single WI (or anything else that isn't true for all the items of one of the ballot groups).
```suggestion
          bool AnyResult = sycl::any_of_group(BallotGroup, OriginalLID == 0);
```

Pennycook replied:
Fixed in 7359d6b9797a.  I used `LID` instead of `OriginalLID` because we still want the condition to be true on at least one work-item in each partition.

Pennycook replied:
I updated the predicate but forgot to update the rest of the test, which is why it started failing.  Fixed in 354179320cdd.

aelovikov-intel commented on sycl/test-e2e/NonUniformGroups/ballot_group_algorithms.cpp:
```diff
@@ -0,0 +1,131 @@
+          uint32_t ReduceResult =
+              sycl::reduce_over_group(BallotGroup, 1, sycl::plus<>());
+          ReduceAcc[WI] =
+              (ReduceResult == BallotGroup.get_local_linear_range());
```
Strictly speaking, the previous test only verified `get_local_range()` and not `get_local_linear_range()` but we can leave this to CTS.

aelovikov-intel commented on sycl/test-e2e/NonUniformGroups/is_user_constructed.cpp:
```diff
@@ -0,0 +1,14 @@
+// RUN: %clangxx -fsycl -fsyntax-only -fsycl-targets=%sycl_triple %s -o %t.out
+
+#include <sycl/sycl.hpp>
+namespace syclex = sycl::ext::oneapi::experimental;
+
+static_assert(
+    syclex::is_user_constructed_group_v<syclex::ballot_group<sycl::sub_group>>);
+static_assert(syclex::is_user_constructed_group_v<
+              syclex::cluster_group<1, sycl::sub_group>>);
+static_assert(syclex::is_user_constructed_group_v<
+              syclex::cluster_group<2, sycl::sub_group>>);
+static_assert(
+    syclex::is_user_constructed_group_v<syclex::tangle_group<sycl::sub_group>>);
+static_assert(syclex::is_user_constructed_group_v<syclex::opportunistic_group>);
```
I'd slightly prefer to have this fused with the previous one but won't insist on that.

aelovikov-intel commented on sycl/test-e2e/NonUniformGroups/opportunistic_group.cpp:
```diff
@@ -0,0 +1,68 @@
+                syclex::this_kernel::get_opportunistic_group();
+
+            bool Match = true;
+            Match &= (OpportunisticGroup.get_group_id() == 0);
+            Match &= (OpportunisticGroup.get_local_id() <
+                      OpportunisticGroup.get_local_range());
+            Match &= (OpportunisticGroup.get_group_range() == 1);
+            Match &= (OpportunisticGroup.get_local_linear_range() <=
+                      SG.get_local_linear_range());
```
I'd suggest writing all the ranges/WIs and verifying their sum/existence on the host.

Pennycook replied:
In some cases, that requires a lot more data to be sent back to the host, though.  Doing the test inside of the kernel means we have access to all the values of `nd_item`, `sub_group` and the `opportunistic_group`.  Getting all those values on the host would require a bunch of additional accessors.

aelovikov-intel commented on sycl/test-e2e/NonUniformGroups/tangle_group.cpp:
```diff
@@ -0,0 +1,69 @@
+          auto WI = item.get_global_id();
+          auto SG = item.get_sub_group();
+
+          // Split into odd and even work-items via control flow
+          // Branches deliberately duplicated to test impact of optimizations
+          // This only reliably works with optimizations disabled right now
```
```suggestion
          // Split into odd and even work-items via control flow.
          // Branches deliberately duplicated to test impact of optimizations.
          // This only reliably works with optimizations disabled right now.
```

Pennycook commented:
All the tests passed, so I'll work on making some of these formatting changes tomorrow.

One thing to possibly look at on the CI side: this got stuck waiting on "Stop AWS" for a really long time -- several hours after all the test suite runs had completed.  I'm not sure what that action does, but it struck me as odd that it didn't get scheduled sooner.

aelovikov-intel commented:
> this got stuck waiting on "Stop AWS" for a really long time

That seems to be the issue with Github's public runners that we use for the ultra-lightweight tasks. We've been hitting this issue elsewhere recently too.

Pennycook commented on sycl/include/sycl/detail/spirv.hpp:
```diff
@@ -94,13 +103,37 @@ void GenericCall(const Functor &ApplyToBytes) {
 }
 
-template <typename Group> bool GroupAll(bool pred) {
+template <typename Group> bool GroupAll(Group g, bool pred) {
```
Good catch.  Fixed in ee1f3a300a1825dae3241de0176ee67fc50e3169.

Pennycook commented on sycl/test-e2e/NonUniformGroups/ballot_group_algorithms.cpp:
```diff
@@ -0,0 +1,131 @@
+              Visible += TmpAcc[Other];
+            }
+          }
+          BarrierAcc[WI] = Visible;
```
Fixed in 9df2196e1fb9.

JackAKirk commented:
Hi @Pennycook 

I'm in the middle of implementing cuda support for these algorithms on top of your implementation and I'm at the point where it would be good to ask for your feedback on a few small implementation issues. I've also implemented cuda support for `cluster_group` and `opportunistic_group` on top of your implementations in the commit I reference here.
Note that here:
https://github.com/intel/llvm/commit/c546762ca2382a99b76a0d0e8f5e39db0289d092#diff-23a47145f432e0af1de65698d91f80a6d0c61c2dcfa6db8a8d97ef84b19dbf59R124
I've added a new friend function to ballot_group `GetMask`, and I've correspondingly removed the friend status from `sycl::detail::IdToMaskPosition` since it can simply call `GetMask`. This function is required for at least `ballot_group` and `opportunistic_group` which already have a `SubGroupMask` member that is required in the NVPTX backend for the different supported group algorithms. This leads me to the questions: 

The NVPTX backend could implement `cluster_group` without introducing a `SubGroupMask Mask` member, but at the moment I have added the Mask only for NVPTX here: https://github.com/intel/llvm/commit/c546762ca2382a99b76a0d0e8f5e39db0289d092#diff-9f01b860115d8a8dab22444aefe9db812b7d11ef489d35a6e2470340436f098bR115
The advantage to storing the mask is that it would otherwise need to be recreated for every non_uniform group algorithm call taking a `cluster_group` in the NVPTX backend via `cluster_group` methods. Do you prefer the mask storing approach or the mask recreation approach?
If the mask storing approach is preferable then `GetMask` would obviously additionally need to support the `cluster_group` case in the NVPTX backend. Currently I notice that your implementation of `sycl::detail::IdToMaskPosition` supports all Non-Uniform groups, rather than just `ballot_group` and `opportunistic_group`, but I guessed that this was an oversight? This observation made me want to check that in the mask storing approach: would you prefer to wrap `GetMask` for `cluster_group` and the `SubGroupMask` member of `cluster_group` in `#if defined(NVPTX)`, or just support them for Intel backends too but have them unused?

Other than this I don't think there are any issues. I have implemented GroupAny, GroupAll, GroupBarrier, and GroupBroadcast for `ballot_group` as you can see, and all these algorithms should call identical code for all NVPTX non uniform groups that store a mask. I've been testing them with the tests that you already wrote. Reduce/(inc/exc)Scan are a bit more complex but I do not believe they introduce any other implementation questions.
Let me know if you have any other thoughts/questions on the NVPTX implemention so far.

Thanks

Pennycook commented:
> I'm in the middle of implementing cuda support for these algorithms on top of your implementation and I'm at the point where it would be good to ask for your feedback on a few small implementation issues.

Sorry for the delayed response, @JackAKirk.

> I've also implemented cuda support for `cluster_group` and `opportunistic_group` on top of your implementations in the commit I reference here. Note that here: [c546762#diff-23a47145f432e0af1de65698d91f80a6d0c61c2dcfa6db8a8d97ef84b19dbf59R124](https://github.com/intel/llvm/commit/c546762ca2382a99b76a0d0e8f5e39db0289d092#diff-23a47145f432e0af1de65698d91f80a6d0c61c2dcfa6db8a8d97ef84b19dbf59R124) I've added a new friend function to ballot_group `GetMask`, and I've correspondingly removed the friend status from `sycl::detail::IdToMaskPosition` since it can simply call `GetMask`. This function is required for at least `ballot_group` and `opportunistic_group` which already have a `SubGroupMask` member that is required in the NVPTX backend for the different supported group algorithms.

This is a good idea.

> The NVPTX backend could implement `cluster_group` without introducing a `SubGroupMask Mask` member, but at the moment I have added the Mask only for NVPTX here: [c546762#diff-9f01b860115d8a8dab22444aefe9db812b7d11ef489d35a6e2470340436f098bR115](https://github.com/intel/llvm/commit/c546762ca2382a99b76a0d0e8f5e39db0289d092#diff-9f01b860115d8a8dab22444aefe9db812b7d11ef489d35a6e2470340436f098bR115) The advantage to storing the mask is that it would otherwise need to be recreated for every non_uniform group algorithm call taking a `cluster_group` in the NVPTX backend via `cluster_group` methods. Do you prefer the mask storing approach or the mask recreation approach?

Hm.  I don't think I have a strong preference, because it's not immediately obvious to me which the compiler is going to be better at optimizing.  My gut says that storing the mask might be slightly easier to optimize: it's unlikely that somebody would create a `cluster_group` and not use it--so we don't have to worry about the compiler's ability to optimize away the mask logic--and I trust the compiler to avoid spilling/filling the calculated mask as best it can. I'm less confident that the compiler could reliably optimize away repeated mask calculations.

> Currently I notice that your implementation of `sycl::detail::IdToMaskPosition` supports all Non-Uniform groups, rather than just `ballot_group` and `opportunistic_group`, but I guessed that this was an oversight? This observation made me want to check that in the mask storing approach: would you prefer to wrap `GetMask` for `cluster_group` and the `SubGroupMask` member of `cluster_group` in `#if defined(NVPTX)`, or just support them for Intel backends too but have them unused?

Honestly, I'm not sure.  I think you're right that it would make sense for `IdToMaskPosition` to only support the non-uniform groups that actually store masks.  Perhaps this should be a more general, and not mention masks?  The other groups may need to perform similar operations (e.g. in the case of a `cluster_group`, we probably need to do some modular arithmetic somewhere).

> Other than this I don't think there are any issues. I have implemented GroupAny, GroupAll, GroupBarrier, and GroupBroadcast for `ballot_group` as you can see, and all these algorithms should call identical code for all NVPTX non uniform groups that store a mask. I've been testing them with the tests that you already wrote. Reduce/(inc/exc)Scan are a bit more complex but I do not believe they introduce any other implementation questions. Let me know if you have any other thoughts/questions on the NVPTX implemention so far.

The only thing I'm curious about is this:
```c++
res[0] = __nvvm_vote_ballot_sync(threads, predicate); // couldnt call this within intel impl because undefined behaviour if not all reach it?
```

I understand the comment, I think, -- `threads` is from `__clc_membermask()`, which always returns a mask representing the entire warp.  But it's not obvious to me where this is being used by the new groups.  Have I missed it?

---

Thank you again for working on this, I really appreciate it.  Now that I'm back from vacation, I'll renew my efforts to get this fixed and merged in, along with the other group types.

againull commented:
Hi @Pennycook, could you please fix post-commit issues for this PR. They are mostly werror problems: 
https://github.com/intel/llvm/commit/70a35de4eacbf6bf1d7b4d2d706a235ed066f794

aelovikov-intel commented:
I also see this https://github.com/intel/llvm/actions/runs/4767700304/jobs/8476298480#step:7:3089 :
```
D:\github\_work\llvm\llvm\build\include\sycl/ext/oneapi/experimental/non_uniform_groups.hpp(57): error C3861: '__builtin_unreachable': identifier not found
D:\github\_work\llvm\llvm\build\include\sycl/ext/oneapi/experimental/non_uniform_groups.hpp(57): note: '__builtin_unreachable': function declaration must be available as none of the arguments depend on a template parameter
```

I'm not 100% sure it's caused by this, but looks so.

JackAKirk commented:
> 
> I understand the comment, I think, -- `threads` is from `__clc_membermask()`, which always returns a mask representing the entire warp. But it's not obvious to me where this is being used by the new groups. Have I missed it?
> 
> Thank you again for working on this, I really appreciate it. Now that I'm back from vacation, I'll renew my efforts to get this fixed and merged in, along with the other group types.

Sorry for the confusion of the comment. This was really a note to myself when I was considering whether I could make a cuda impl that reused the same spirv functions that do not take a mask. The relevance is simply that you can use `__nvvm_vote_ballot_sync(FULLMASK, predicate);` to return the mask for the `ballot_group`.

All sounds good. I also think it is best to store a mask for `fixed_size_group` for cuda backend, which I have done in https://github.com/intel/llvm/pull/9182 (see this PR for more details on cuda backend impl status). I opened this draft PR mainly for your convenience of seeing the requirements from the cuda backend implementation. I think I've now implemented enough so that there should be no more surprises with cuda backend requirements. Really the only missing thing is the scans implementations which will follow a very similar pattern to the reduce_over_group implementation, except it is simpler because there is no special sm_80 version.
I will wait to see how you implement the `opportunistic_group` algorithms before finalizing the design. In the mean-time I mainly just need to implement this case (https://github.com/intel/llvm/pull/9182/files#diff-185200e89425bd69d2e35ca3ba250d40934230436b718d4f2eecbab9325cc972R170) which is quite involved.
